### ISSUE TITLE
docker keeps saying that container is 'unhealty'

### ISSUE THREAD

**Issue by stempst0r** [haugene/docker-transmission-openvpn#1582]
I'm running the container on a raspberry pi 2 with hyperiotos.
Everything is running fine I guess, but docker says the container is unhealty.

```
$ docker ps
CONTAINER ID        IMAGE                          COMMAND                  CREATED             STATUS                      PORTS                                                                                                  NAMES
409142cbefe3        haugene/transmission-openvpn   "dumb-init /etc/open…"   10 minutes ago      Up 10 minutes (unhealthy)   0.0.0.0:9091->9091/tcp                                                                                 transmission-openvpn
```

I also checked if the VPN is running. Indeed with `curl ifconfig.io` from inside the container I get another IP than from outside the container, so VPN is working.

**docker-compose**
```
$ cat docker-compose.yml
version: '3.3'
services:
    transmission-openvpn:
        cap_add:
            - NET_ADMIN
        volumes:
            - '/mnt/nas/torrent:/data'
        environment:
            - OPENVPN_PROVIDER=VPNUNLIMITED
            - OPENVPN_CONFIG=fr,lu,ro
            - OPENVPN_USERNAME=***
            - OPENVPN_PASSWORD=***
            - LOCAL_NETWORK=192.168.188.0/24
        dns:
            - 192.168.188.5
            - 1.1.1.1
        logging:
            driver: json-file
            options:
                max-size: 10m
        ports:
            - '9091:9091'
        image: haugene/transmission-openvpn
        restart: on-failure
        container_name: transmission-openvpn
```

**Logs**
```
$ docker logs 409142cbefe3
Starting container with revision: 42d00652323b7b7fbee1689a54219599f4795a04
Creating TUN device /dev/net/tun
Using OpenVPN provider: VPNUNLIMITED
3 servers found in OPENVPN_CONFIG, ro chosen randomly
Starting OpenVPN using config ro.ovpn
Setting OpenVPN credentials...
adding route to local network 192.168.188.0/24 via 172.22.0.1 dev eth0
Wed Dec  2 09:42:11 2020 OpenVPN 2.4.9 armv7-alpine-linux-musleabihf [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Apr 20 2020
Wed Dec  2 09:42:11 2020 library versions: OpenSSL 1.1.1g  21 Apr 2020, LZO 2.10
Wed Dec  2 09:42:11 2020 NOTE: the current --script-security setting may allow this configuration to call user-defined scripts
Wed Dec  2 09:42:15 2020 TCP/UDP: Preserving recently used remote address: [AF_INET]185.144.83.13:1194
Wed Dec  2 09:42:15 2020 UDP link local: (not bound)
Wed Dec  2 09:42:15 2020 UDP link remote: [AF_INET]185.144.83.13:1194
Wed Dec  2 09:42:16 2020 WARNING: 'link-mtu' is used inconsistently, local='link-mtu 1542', remote='link-mtu 1602'
Wed Dec  2 09:42:16 2020 WARNING: 'cipher' is used inconsistently, local='cipher BF-CBC', remote='cipher AES-256-CBC'
Wed Dec  2 09:42:16 2020 WARNING: 'auth' is used inconsistently, local='auth SHA1', remote='auth SHA512'
Wed Dec  2 09:42:16 2020 WARNING: 'keysize' is used inconsistently, local='keysize 128', remote='keysize 256'
Wed Dec  2 09:42:16 2020 [openvpn2.vpnunlimitedapp.com] Peer Connection Initiated with [AF_INET]185.144.83.13:1194
Wed Dec  2 09:42:22 2020 TUN/TAP device tun0 opened
Wed Dec  2 09:42:22 2020 /sbin/ip link set dev tun0 up mtu 1500
Wed Dec  2 09:42:22 2020 /sbin/ip addr add dev tun0 local 10.200.0.74 peer 10.200.0.73
Wed Dec  2 09:42:22 2020 /etc/openvpn/tunnelUp.sh tun0 1500 1553 10.200.0.74 10.200.0.73 init
Up script executed with tun0 1500 1553 10.200.0.74 10.200.0.73 init
Updating TRANSMISSION_BIND_ADDRESS_IPV4 to the ip of tun0 : 10.200.0.74
Updating Transmission settings.json with values from env variables
Using existing settings.json for Transmission /data/transmission-home/settings.json
Overriding bind-address-ipv4 because TRANSMISSION_BIND_ADDRESS_IPV4 is set to 10.200.0.74
Overriding download-dir because TRANSMISSION_DOWNLOAD_DIR is set to /data/completed
Overriding incomplete-dir because TRANSMISSION_INCOMPLETE_DIR is set to /data/incomplete
Overriding rpc-port because TRANSMISSION_RPC_PORT is set to 9091
Overriding watch-dir because TRANSMISSION_WATCH_DIR is set to /data/watch
sed'ing True to true

-------------------------------------
Transmission will run as
-------------------------------------
User name:   root
User uid:    0
User gid:    0
-------------------------------------

STARTING TRANSMISSION
Transmission startup script complete.
Wed Dec  2 09:42:23 2020 Initialization Sequence Completed
```

**Host system:**
`$ uname -a
Linux black-pearl 5.4.72-v7+ #1356 SMP Thu Oct 22 13:56:54 BST 2020 armv7l GNU/Linux`

```
$ docker -v
Docker version 19.03.14, build 5eb3275d40

```

Hope I provided everything needed, thanks for the awsome container!

**Comment by haugene** [comment]
Hmm. It will say that the container is unhealthy as long as the health check script is failing. That doesn't necessarily mean that it doesn't work, but it would be interesting to see what fails in the script. Thank you for a very well specified issue, it's nice to come to one of these compared to "it doesn't work" as some issues basically put it :laughing: 

If you do `docker inspect transmission-openvpn` you should find a section with "State" - > "Health". Can you post that?

**Comment by stempst0r** [comment]
First thanks for your response! Didn't have access to the machine for the last days, because of that the delay. Here is the output of `docker inspect transmission-openvpn`

```
"Health": {
                "Status": "unhealthy",
                "FailingStreak": 2420,
                "Log": [
                    {
                        "Start": "2020-12-06T22:24:53.614170248Z",
                        "End": "2020-12-06T22:24:59.305009928Z",
                        "ExitCode": 1,
                        "Output": "PING google.com (216.58.214.142): 56 data bytes\n64 bytes from 216.58.214.142: seq=0 ttl=113 time=97.965 ms\n\n--- google.com ping statistics ---\n2 packets transmitted, 1 packets received, 50% packet loss\nround-trip min/avg/max = 97.965/97.965/97.965 ms\nNetwork is down\n"
                    },
                    {
                        "Start": "2020-12-06T22:25:59.348762669Z",
                        "End": "2020-12-06T22:26:05.00527991Z",
                        "ExitCode": 1,
                        "Output": "PING google.com (172.217.17.238): 56 data bytes\n64 bytes from 172.217.17.238: seq=0 ttl=113 time=89.376 ms\n\n--- google.com ping statistics ---\n2 packets transmitted, 1 packets received, 50% packet loss\nround-trip min/avg/max = 89.376/89.376/89.376 ms\nNetwork is down\n"
                    },
                    {
                        "Start": "2020-12-06T22:27:05.045131429Z",
                        "End": "2020-12-06T22:27:10.693205139Z",
                        "ExitCode": 1,
                        "Output": "PING google.com (172.217.17.238): 56 data bytes\n64 bytes from 172.217.17.238: seq=0 ttl=113 time=89.192 ms\n\n--- google.com ping statistics ---\n2 packets transmitted, 1 packets received, 50% packet loss\nround-trip min/avg/max = 89.192/89.192/89.192 ms\nNetwork is down\n"
                    },
                    {
                        "Start": "2020-12-06T22:28:10.729377572Z",
                        "End": "2020-12-06T22:28:16.389096899Z",
                        "ExitCode": 1,
                        "Output": "PING google.com (172.217.17.238): 56 data bytes\n64 bytes from 172.217.17.238: seq=0 ttl=113 time=89.678 ms\n\n--- google.com ping statistics ---\n2 packets transmitted, 1 packets received, 50% packet loss\nround-trip min/avg/max = 89.678/89.678/89.678 ms\nNetwork is down\n"
                    },
                    {
                        "Start": "2020-12-06T22:29:16.432009562Z",
                        "End": "2020-12-06T22:29:22.053042577Z",
                        "ExitCode": 1,
                        "Output": "PING google.com (172.217.17.238): 56 data bytes\n64 bytes from 172.217.17.238: seq=0 ttl=113 time=89.960 ms\n\n--- google.com ping statistics ---\n2 packets transmitted, 1 packets received, 50% packet loss\nround-trip min/avg/max = 89.960/89.960/89.960 ms\nNetwork is down\n"
                    }
                ]
```

For me, it sounds simply like a problem with an unstable connection because there is packetloss? The Pi is wired, so in my private network the connection should be fine. I have a Pi-Hole-instance as local DNS, maybe this blocks some addresses? I set the DNS of the docker-container to "1.1.1.1" and "1.0.0.1" (both cloudflare-DNS-server) and restarted the container, maybe the problem is solved with that?

**Comment by stempst0r** [comment]
Actually changing the DNS doesn't solve the issue.

**Comment by clement-z** [comment]
Could you try:

* To manually ping google.com (without `-c 2` as in the healthcheck script)
* Same with other target hosts
* Same from host OS

I'm interested to see if maybe there is a pattern for the ping fails (for example, maybe only the first is ever blocked ?)

---

Note: your public IP (of 5 days ago) seems to be appearing in the first log message. You might want to redact it.

**Comment by clement-z** [comment]
Hi @stempst0r, I was able to reproduce the issue manually on my container. I'm pretty sure it is related to the `-w` switch not doing what we think it does (and what the doc says?). `busybox` provides the `timeout` which we can use to replace it if needed. I will look into it and open a PR.

**Comment by stempst0r** [comment]
> 
> 
> Hi @stempst0r, I was able to reproduce the issue manually on my container. I'm pretty sure it is related to the `-w` switch not doing what we think it does (and what the doc says?). `busybox` provides the `timeout` which we can use to replace it if needed. I will look into it and open a PR.

Thanks for your time! 
Docker is quite new to me and it's a hard time understanding the virtual networking for me. If I can help with any further information of my error, let me know.

**Comment by clement-z** [comment]
Hmm I cannot seem to reproduce it anymore. And `-w` does seem to work as it should... I don't really know what changed.

@stempst0r is the issue still present on your side ?

**Comment by clement-z** [comment]
And if it is, can you try running (in order until it works):

* `docker exec container_name ping -c 2 -w 5 google.com`
* `docker exec container_name ping -c 2 -w 10 google.com`
* `docker exec container_name timeout 5 ping -c 2 google.com`

For me all of these work fine now, but the first one gave me the same 50% loss error last time I tried, even though 5 seconds hadn't elapsed... On my side nothing should have changed network wise.

**Comment by stempst0r** [comment]
> 
> 
> And if it is, can you try running (in order until it works):
> 
>     * `docker exec container_name ping -c 2 -w 5 google.com`
> 
>     * `docker exec container_name ping -c 2 -w 10 google.com`
> 
>     * `docker exec container_name timeout 5 ping -c 2 google.com`
> 
> 
> For me all of these work fine now, but the first one gave me the same 50% loss error last time I tried, even though 5 seconds hadn't elapsed... On my side nothing should have changed network wise.

Here are my results:

```
$ docker exec transmission-openvpn ping -c 2 -w 5 google.com
PING google.com (172.217.17.142): 56 data bytes
64 bytes from 172.217.17.142: seq=0 ttl=118 time=45.912 ms

--- google.com ping statistics ---
2 packets transmitted, 1 packets received, 50% packet loss
round-trip min/avg/max = 45.912/45.912/45.912 ms
```

```
$ docker exec transmission-openvpn ping -c 2 -w 10 google.com
PING google.com (216.58.214.14): 56 data bytes
64 bytes from 216.58.214.14: seq=0 ttl=118 time=56.604 ms
64 bytes from 216.58.214.14: seq=1 ttl=118 time=40.971 ms

--- google.com ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 40.971/48.787/56.604 ms
```

```
$ docker exec transmission-openvpn timeout 5 ping -c 2 google.com
PING google.com (216.58.214.14): 56 data bytes
64 bytes from 216.58.214.14: seq=0 ttl=118 time=105.033 ms
```

Hope this info can help fix the bug. Tbh idk in which file I should change w 5 to w 10 so I could fix it myself. 
Thanks for your efforts!

**Comment by stempst0r** [comment]
Ok, with your tips I was able to fix it for me. Did the following:

I edited the healthcheck.sh script inside the container /etc/scripts/healtcheck.sh. Changed line 14 from
`ping -c 2 -w 5 $HOST # Get at least 2 responses and timeout after 5 seconds`
to
`ping -c 2 -w 10 $HOST # Get at least 2 responses and timeout after 5 seconds`

Afterwards I restared the container via `docker-compose restart` and now the container is listed as healthy. 
Basically made a small crashcourse how docker container works and how to get things edited via 'vi' 🤣 

Don't know if this value should be changed globally? Seems like this issue only exists with VPNUnlimited as provider?

Anyway thanks for all the effort. I'll close the issue with this comment :)

**Comment by clement-z** [comment]
> Don't know if this value should be changed globally? Seems like this issue only exists with VPNUnlimited as provider?

I am using PIA, and I still had (at some point) the same error as you had. The timeout is just there so that we don't wait indefinitely for the ping. Bumping its duration should have no incidence on the functioning of the container and the ability to say that the connection has dropped.

Note that if you ever use a new container (after a new `docker run`) you will have to re-do this edit. I already had this patch ready [in one of my branches](https://github.com/clement-z/docker-transmission-openvpn/commit/ca7ddfa6061f111d2e7c02bbc859b036a9c42f93), but since I couldn't reproduce the issue anymore, I was waiting for your reply (to see if going to `-w 10` fixed your issue) before asking for merge here.
